tools/mo/openvino/tools/mo/moc_frontend/pipeline.py

```diff
@@ -136,17 +136,50 @@ def create_target_input_shapes(new_input_places):
 
     if freeze_placeholder:
         for name, value in freeze_placeholder.items():
-            for node in user_shapes:
-                if node.get('input_name') == name:
-                    place = node['node']
-                    if node.get('shape'):
-                        input_model.set_partial_shape(place, node['shape'])
-                    if node.get('data_type'):
-                        value = np.array(value, dtype=node['data_type'])
-                        input_model.set_element_type(place, Type(node['data_type']))
-                    else:
-                        value = np.array(value, dtype=np.float32)
-                    input_model.set_tensor_value(place, value)
+            node = None
+            # look for the certain place in user_shapes
+            for node_cur in user_shapes:
+                if node_cur.get('input_name') == name:
+                    node = node_cur
+                    break
+            if node is None:
+                raise Error("Please check correctness of the command-line. "
+                            "Place (operation or tensor) with name {} is not found.".format(name))
+            place = node.get('node')
+
+            if node.get('shape'):
+                input_model.set_partial_shape(place, node['shape'])
+
+            if node.get('data_type'):
+                dtype = node['data_type']
+                ov_type = Type(dtype)
+            else:
+                # we need to detect type of Placeholder
+                try:
+                    ov_type = input_model.get_element_type(place)
+                except NotImplementedFailure:
+                    raise Error("Please specify type for value freezing {} node explicitly "
+                                "because the frontend does not support automatic type detection.".format(name))
+                # in case of cutting graph (or using custom inputs) and unspecified type,
+                # the default type is fp32
+                if ov_type == Type.undefined:
+                    ov_type = Type.f32
+                dtype = get_numpy_ctype(ov_type)
+
+            input_model.set_element_type(place, ov_type)
+            # prepare and cast value to dtype
+            from openvino.tools.mo.front.tf.common import tf_data_type_cast
+            from openvino.tools.mo.front.common.partial_infer.utils import mo_array
+            if isinstance(value, list):
+                casted_list = list()
+                for v in mo_array(value):
+                    casted_list.append(tf_data_type_cast[dtype](v))
+                value = mo_array(casted_list, dtype=dtype)
+            else:
+                value = tf_data_type_cast[dtype](value)
```

Comment:
Shouldn't this `tf_data_type_cast` be applied only for TF frontend?

Reply:
No. I should rather rename this method. It is artifact remained in the legacy.

Reply:
Done.